File: python/src/apx_agent/_tool_publish.py
"""publish_tools_to_uc — sync @tool(uc=...) functions into Unity Catalog.

Walks an agent tree, finds every tool annotated with ``@tool(uc=...)``, and
registers each as a Unity Catalog Python function via ``unitycatalog-ai``.
Applies declarative ``grant=[...]`` lists at the same time.

After publishing, the tools are reachable by any UC-aware client (Genie,
Managed MCP, other agents, AI Playground), governed by standard UC
permissions, and auditable through UC's normal lineage and access logs.
The agent itself still executes the Python in-process at runtime — the UC
function is the discovery / external-composition surface, not the
execution path during the LLM loop.

Requires the ``uc`` extra::

    pip install 'apx-agent[uc]'

This pulls in ``unitycatalog-ai``, which provides the
``DatabricksFunctionClient.create_python_function`` primitive used here.
Grants go through the Databricks SDK's permissions API directly.

Usage::

    from apx_agent import Agent, log_agent, publish_tools_to_uc, tool

    @tool(uc="main.agents.tools.classify_intent", grant=["agent_consumers"])
    def classify_intent(query: str) -> str:
        \"\"\"Classify a customer query as billing/technical/account/other.\"\"\"
        return "billing" if "bill" in query.lower() else "other"

    agent = Agent(tools=[classify_intent], instructions="...")

    publish_tools_to_uc(agent)            # registers + grants
    log_agent(agent, model="...", registered_model_name="main.agents.triage")
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from ._resources import _iter_tool_fns
from ._tool import ToolMetadata, get_tool_metadata

if TYPE_CHECKING:
    from databricks.sdk import WorkspaceClient

    from ._agents import BaseAgent

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class PublishResult:
    """Outcome of publishing a single ``@tool(uc=...)`` to Unity Catalog."""

    uc_name: str
    function_name: str
    grants_applied: tuple[str, ...]
    skipped: bool = False
    skip_reason: str | None = None


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _load_uc_client(client: Any | None) -> Any:
    """Return a ``unitycatalog-ai`` DatabricksFunctionClient instance."""
    if client is not None:
        return client
    try:
        # The OSS Unity Catalog AI package layout (unitycatalog/unitycatalog-ai)
        from unitycatalog.ai.core.databricks import DatabricksFunctionClient
    except ImportError as e:  # pragma: no cover — exercised only without extra
        raise ImportError(
            "publish_tools_to_uc requires the unitycatalog-ai package. "
            "Install with: pip install 'apx-agent[uc]'"
        ) from e
    return DatabricksFunctionClient()


def _grant_execute(
    ws: "WorkspaceClient",
    uc_name: str,
    principal: str,
) -> None:
    """Apply ``GRANT EXECUTE ON FUNCTION <uc_name> TO <principal>``.

    Uses the Unity Catalog permissions API directly. Idempotent — the API
    is additive when EXECUTE is already present.
    """
    body = {
        "changes": [
            {
                "principal": principal,
                "add": ["EXECUTE"],
            }
        ]
    }
    # The UC permissions API takes the securable type + full name in the URL
    # and a list of permission changes in the body.
    ws.api_client.do(
        "PATCH",
        f"/api/2.1/unity-catalog/permissions/function/{uc_name}",
        body=body,
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def publish_tools_to_uc(
    agent: "BaseAgent",
    *,
    client: Any | None = None,
    workspace_client: "WorkspaceClient | None" = None,
    replace: bool = True,
    dry_run: bool = False,
) -> list[PublishResult]:
    """Publish every ``@tool(uc=...)`` in the agent tree to Unity Catalog.

    Args:
        agent: The apx-agent ``BaseAgent`` to walk. All nested agents
            (Sequential, Parallel, Loop, Router, Handoff) are visited.
        client: Optional ``DatabricksFunctionClient`` instance to use. When
            omitted, a default client is constructed via
            ``unitycatalog.ai.core.databricks.DatabricksFunctionClient()``.
            Pass an explicit client when you need a non-default profile or
            workspace.
        workspace_client: Optional ``databricks.sdk.WorkspaceClient`` used
            for the GRANT step. When omitted, a default client is built
            from the ambient environment (which is the same auth chain
            the OBO / SP fallback uses elsewhere in apx-agent).
        replace: When ``True`` (default), pass ``replace=True`` to
            ``create_python_function`` so existing functions with the same
            name are overwritten. Set to ``False`` for a "fail if exists"
            mode useful in CI.
        dry_run: When ``True``, walks the tree and reports what would be
            published without making any UC writes. Useful for previewing
            in CI or local dev.

    Returns:
        One ``PublishResult`` per ``@tool(uc=...)`` found in the tree,
        including any that were skipped (and why).
    """
    results: list[PublishResult] = []

    seen: set[str] = set()
    targets: list[tuple[Any, ToolMetadata]] = []
    for fn in _iter_tool_fns(agent):
        meta = get_tool_metadata(fn)
        if meta is None or meta.uc_name is None:
            continue
        if meta.uc_name in seen:
            # Same UC name registered by multiple references to the same
            # tool — publish once.
            continue
        seen.add(meta.uc_name)
        targets.append((fn, meta))

    if not targets:
        return results

    if dry_run:
        for fn, meta in targets:
            results.append(
                PublishResult(
                    uc_name=meta.uc_name or "",
                    function_name=fn.__name__,
                    grants_applied=meta.grants,
                    skipped=True,
                    skip_reason="dry_run",
                )
            )
        return results

    uc_client = _load_uc_client(client)
    ws = workspace_client
    if ws is None and any(meta.grants for _, meta in targets):
        # Only build a default WorkspaceClient if we actually need one for
        # grants — keeps the no-grants path SDK-free.
        from databricks.sdk import WorkspaceClient as _WS
        ws = _WS()

    for fn, meta in targets:
        assert meta.uc_name is not None  # narrowed by the filter above
        catalog, schema, _function = meta.uc_name.split(".")
        try:
            uc_client.create_python_function(
                func=fn,
                catalog=catalog,
                schema=schema,
                replace=replace,
            )
        except Exception as e:
            logger.exception("Failed to publish %s to UC", meta.uc_name)
            raise RuntimeError(
                f"Failed to publish {meta.uc_name!r} to UC: {e}"
            ) from e

        grants_applied: list[str] = []
        for principal in meta.grants:
            try:
                assert ws is not None
                _grant_execute(ws, meta.uc_name, principal)
                grants_applied.append(principal)
            except Exception as e:
                logger.warning(
                    "GRANT EXECUTE on %s to %s failed: %s — "
                    "function is published but ungranted.",
                    meta.uc_name, principal, e,
                )

        results.append(
            PublishResult(
                uc_name=meta.uc_name,
                function_name=fn.__name__,
                grants_applied=tuple(grants_applied),
            )
        )
        logger.info(
            "Published %s to UC (grants: %s)",
            meta.uc_name, ", ".join(grants_applied) or "none",
        )

    return results

File: typescript/src/tool-publish.ts
/**
 * `publishToolsToUc` — sync `defineUcTool` functions into Unity Catalog.
 *
 * TypeScript counterpart to ``python/src/apx_agent/_tool_publish.py``. Walks
 * a list of {@link AgentTool}s, finds every tool annotated with
 * {@link ToolMetadata.ucName}, and registers each as a Unity Catalog Python
 * function via the injected `UcFunctionClient`. Applies declarative grants
 * through the injected `WsClient`.
 *
 * Design differences from the Python version:
 *
 *   - **No agent walking.** Python walks a `BaseAgent` tree via
 *     `_iter_tool_fns(agent)` because Python tools are bare callables held
 *     inside agent objects. In TS, tools are `AgentTool` records constructed
 *     up front; the caller already has a list of them. `publishToolsToUc`
 *     takes `tools: AgentTool[]` directly. If/when a TS agent class hierarchy
 *     materialises and you need tree-walking, build a `collectTools(agent)`
 *     helper next to `collectResourceSpecs` in `resources.ts` and call it
 *     before `publishToolsToUc`.
 *   - **Clients are injected, not constructed.** Python falls back to a
 *     default `DatabricksFunctionClient()` and `WorkspaceClient()`. TS has
 *     no canonical "default Databricks SDK in this process" pattern, so the
 *     caller passes minimal client objects. The interfaces below describe
 *     the exact surface used; any adapter that conforms is acceptable.
 *
 * Python semantics preserved:
 *
 *   - Idempotent enumeration — same `ucName` across multiple tool entries
 *     yields one publish call.
 *   - `dryRun: true` skips all writes and returns results with
 *     `skipped: true, skipReason: 'dry_run'`.
 *   - `replace: true` (default) forwards to `createPythonFunction`.
 *   - `createPythonFunction` failure raises (TS throws an Error wrapping
 *     the underlying cause; Python raised `RuntimeError`).
 *   - Grant failures log-and-continue — the function is published, the
 *     failed principal is dropped from `grantsApplied`.
 */

import { z } from 'zod';

import type { AgentTool } from './agent/tools.js';
import { getToolMetadata } from './tool.js';

// ---------------------------------------------------------------------------
// Client interfaces (injection points)
// ---------------------------------------------------------------------------

/**
 * Minimal surface required from a Unity Catalog function client. Mirrors the
 * `create_python_function` call in `unitycatalog-ai`.
 *
 * Implementations are expected to register `func` as a UC Python function
 * under `catalog`.`schema`; the function name is implementation-defined
 * (typically `func.name`).
 */
export interface UcFunctionClient {
  createPythonFunction(opts: {
    func: unknown;
    catalog: string;
    schema: string;
    replace?: boolean;
  }): Promise<void>;
}

/**
 * Minimal surface required from a Databricks workspace client for the
 * GRANT step. `apiRequest` issues an authenticated REST call against the
 * workspace.
 *
 * Compared to Python's `WorkspaceClient.api_client.do(method, path, body)`,
 * this is a thin async wrapper — any adapter that posts to the Databricks
 * API conforms.
 */
export interface WsClient {
  apiRequest(method: string, path: string, body?: object): Promise<void>;
}

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

/**
 * Outcome of publishing a single `defineUcTool` to Unity Catalog.
 * Frozen at construction.
 */
export interface PublishResult {
  readonly ucName: string;
  readonly functionName: string;
  readonly grantsApplied: readonly string[];
  readonly skipped: boolean;
  readonly skipReason?: string;
}

function makePublishResult(opts: {
  ucName: string;
  functionName: string;
  grantsApplied: readonly string[];
  skipped?: boolean;
  skipReason?: string;
}): PublishResult {
  return Object.freeze({
    ucName: opts.ucName,
    functionName: opts.functionName,
    grantsApplied: Object.freeze([...opts.grantsApplied]),
    skipped: opts.skipped ?? false,
    skipReason: opts.skipReason,
  });
}

// ---------------------------------------------------------------------------
// Options validation
// ---------------------------------------------------------------------------

const PublishOptionsSchema = z.object({
  dryRun: z.boolean().optional(),
  replace: z.boolean().optional(),
});

export interface PublishToolsToUcOptions {
  /**
   * The tools to consider for UC publication. Tools without UC metadata
   * (i.e. not produced by `defineUcTool`) are silently skipped.
   */
  tools: readonly AgentTool[];
  /**
   * Client used to register the Python function in UC.
   */
  ucClient: UcFunctionClient;
  /**
   * Workspace client used for the GRANT step. Required only when at least
   * one tool declares `grant`; omit otherwise.
   */
  workspaceClient?: WsClient;
  /**
   * When `true`, walks the list and reports what would be published without
   * making any UC writes.
   */
  dryRun?: boolean;
  /**
   * When `true` (default), forwards `replace: true` to
   * `createPythonFunction` so existing functions with the same name are
   * overwritten. Set to `false` for a "fail if exists" mode useful in CI.
   */
  replace?: boolean;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async function grantExecute(ws: WsClient, ucName: string, principal: string): Promise<void> {
  await ws.apiRequest('PATCH', `/api/2.1/unity-catalog/permissions/function/${ucName}`, {
    changes: [
      {
        principal,
        add: ['EXECUTE'],
      },
    ],
  });
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Publish every tool with UC metadata to Unity Catalog.
 *
 * Returns one {@link PublishResult} per processed tool. Tools whose
 * `ucName` was already published in this call are deduplicated (first
 * occurrence wins). Grant failures are logged via `console.warn` and the
 * principal is dropped from `grantsApplied`; the publish itself proceeds.
 *
 * Throws if any `createPythonFunction` call fails — that's a hard failure
 * (matches Python's `RuntimeError` semantics).
 */
export async function publishToolsToUc(opts: PublishToolsToUcOptions): Promise<PublishResult[]> {
  PublishOptionsSchema.parse({ dryRun: opts.dryRun, replace: opts.replace });

  const replace = opts.replace ?? true;
  const dryRun = opts.dryRun ?? false;

  const results: PublishResult[] = [];

  // Idempotent enumeration: same ucName seen twice → publish once.
  const seen = new Set<string>();
  const targets: Array<{ tool: AgentTool; ucName: string; grants: readonly string[] }> = [];
  for (const tool of opts.tools) {
    const meta = getToolMetadata(tool);
    if (meta === null || meta.ucName === null) continue;
    if (seen.has(meta.ucName)) continue;
    seen.add(meta.ucName);
    targets.push({ tool, ucName: meta.ucName, grants: meta.grants });
  }

  if (targets.length === 0) return results;

  if (dryRun) {
    for (const t of targets) {
      results.push(
        makePublishResult({
          ucName: t.ucName,
          functionName: t.tool.name,
          grantsApplied: t.grants,
          skipped: true,
          skipReason: 'dry_run',
        }),
      );
    }
    return results;
  }

  const needsGrants = targets.some((t) => t.grants.length > 0);
  if (needsGrants && opts.workspaceClient === undefined) {
    throw new Error(
      'publishToolsToUc: at least one tool declares grants but no workspaceClient was supplied. ' +
        'Pass `workspaceClient` to apply UC permissions.',
    );
  }

  for (const t of targets) {
    const segments = t.ucName.split('.');
    // `defineUcTool` validates this at construction, but guard here too
    // for callers that bypass the factory.
    if (segments.length !== 3) {
      throw new Error(
        `publishToolsToUc: invalid UC name ${JSON.stringify(t.ucName)} — expected catalog.schema.function`,
      );
    }
    const [catalog, schema] = segments as [string, string, string];

    try {
      await opts.ucClient.createPythonFunction({
        func: t.tool,
        catalog,
        schema,
        replace,
      });
    } catch (err) {
      const reason = err instanceof Error ? err.message : String(err);
      throw new Error(`Failed to publish ${JSON.stringify(t.ucName)} to UC: ${reason}`);
    }

    const grantsApplied: string[] = [];
    for (const principal of t.grants) {
      try {
        // `workspaceClient` is guaranteed defined when we reach this branch
        // because we threw above if grants were requested without one.
        await grantExecute(opts.workspaceClient!, t.ucName, principal);
        grantsApplied.push(principal);
      } catch (err) {
        const reason = err instanceof Error ? err.message : String(err);
        // eslint-disable-next-line no-console
        console.warn(
          `GRANT EXECUTE on ${t.ucName} to ${principal} failed: ${reason} — ` +
            `function is published but ungranted.`,
        );
      }
    }

    results.push(
      makePublishResult({
        ucName: t.ucName,
        functionName: t.tool.name,
        grantsApplied,
      }),
    );
  }

  return results;
}
